### ISSUE TITLE
Unicode-escaped ampersand in query string returned by Admin API recovery link endpoint

### ISSUE THREAD

**Issue by fenech** [ory/kratos#1188]
**Describe the bug**

Ampersand `&` escaped to `\u0026` in recovery link returned from [Admin API recovery link endpoint](https://www.ory.sh/kratos/docs/reference/api/#create-a-recovery-link). The escape sequence must be replaced by `&` in order to use the link.

**Reproducing the bug**

1. Run Kratos following Quickstart instructions (SQLite DB can be used)
2. Add a user, making a note of user's ID
3. Send a cURL request to the recovery link endpoint of the admin API. replacing the `identity_id` value with the newly-created user's ID:
    ```sh
    curl -X POST http://localhost:4434/recovery/link -H 'Content-Type: application/json' \
    -d '{"identity_id": "57ff394d-4667-4f5f-95db-db2f6f153741", "expires_in":"12h"}'
4. The JSON response is of the form:
    ```json
    {
      "recovery_link":"http://127.0.0.1:4455/.ory/kratos/public/self-service/recovery/methods/link?flow=ac1ce7ac-76df-481e-84de-9df59ee5d33d\u0026token=OqgpDBnwpPNb36Sk4sRFlVrAkPUtVavF",
      "expires_at":"2021-03-30T21:01:50.10467534Z"
    }
    ```
_Server logs_

```
time=2021-03-30T09:11:52Z level=info msg=started handling request method=POST name=admin#http://kratos:4434/ remote=10.0.2.100:36828 request=/recovery/link
time=2021-03-30T09:11:52Z level=info msg=A recovery link has been created. audience=audit email_address=Value is sensitive and has been redacted. To see the value set config key "log.leak_sensitive_values = true" or environment variable "LOG_LEAK_SENSITIVE_VALUES=true". identity_id=57ff394d-4667-4f5f-95db-db2f6f153741 recovery_link_token=Value is sensitive and has been redacted. To see the value set config key "log.leak_sensitive_values = true" or environment variable "LOG_LEAK_SENSITIVE_VALUES=true". service_name=ORY Kratos service_version=v0.5.4-alpha.1 via=email
time=2021-03-30T09:11:52Z level=info msg=completed handling request method=POST name=admin#http://kratos:4434/ remote=10.0.2.100:36828 request=/recovery/link status=200 text_status=OK took=2.767399ms
```

_Server configuration_

```yaml
session:
  lifespan: 24h
serve:
  public:
    base_url: http://127.0.0.1:4455/.ory/kratos/public
    cors:
      enabled: true
  admin:
    base_url: http://kratos:4434/
selfservice:
  default_browser_return_url: http://127.0.0.1:4455/
  whitelisted_return_urls:
    - http://127.0.0.1:4455
    - http://127.0.0.1:4455/app/services
  methods:
    password:
      enabled: true
  flows:
    error:
      ui_url: http://127.0.0.1:4455/auth/error
    settings:
      ui_url: http://127.0.0.1:4455/user/settings
      privileged_session_max_age: 15m
    registration:
      ui_url: http://127.0.0.1:4455/auth/register
      after:
        password:
          hooks:
            - hook: session
        oidc:
          hooks:
            - hook: session
    recovery:
      enabled: true
      ui_url: http://127.0.0.1:4455/auth/recovery
    logout:
      after:
        default_browser_return_url: http://127.0.0.1:4455/
    login:
      ui_url: http://127.0.0.1:4455/auth/login
      after:
        default_browser_return_url: http://127.0.0.1:4455/app/services
      lifespan: 10m
log:
  level: debug
secrets:
  cookie:
    - PLEASE-CHANGE-ME-I-AM-VERY-INSECURE
hashers:
  argon2:
    parallelism: 1
    memory: 131072
    iterations: 2
    salt_length: 16
    key_length: 16
identity:
  default_schema_url: file:///etc/config/kratos/user.schema.json
  schemas:
    - id: admin
      url: file:///etc/config/kratos/admin.schema.json
courier:
  smtp:
    connection_uri: smtps://test:test@mailslurper:1025/?skip_ssl_verify=true
```

**Expected behavior**

There is an un-escaped `&` between the two query parameters in value of `recovery_link` returned by Kratos.

**Environment**

- Version: oryd/kratos:v0.5.4-alpha.1-sqlite
- Environment: Fedora, Podman

**Additional context**

The response is generated here:

https://github.com/ory/kratos/blob/4bda1f09472bd65608cafe0d5d44287487da6f70/selfservice/strategy/link/strategy_recovery.go#L187

and the value returned by `urlx.CopyWithQuery` contains the ampersand, as expected: https://play.golang.org/p/A97pqt_ZlHO

**Comment by aeneasr** [comment]
Thank you, contributions towards fixing this are highly appreciated!

**Comment by fenech** [comment]
The fix would be to make a change to this function https://github.com/ory/herodot/blob/master/json.go#L64
and instead of using the default JSON marshaller, use something like this:

```go
buf := new(bytes.Buffer)
enc := json.NewEncoder(buf)
enc.SetEscapeHTML(false)
err := enc.Encode(e)
```

But I'm not sure if that's something that you would want to enable globally. Maybe it could be a configurable property on the JSONWriter instance?

**Comment by aeneasr** [comment]
Yeah that should definitely not be enabled globally. I think we should probably have an option to enable it for very specific payloads?

**Comment by fenech** [comment]
To enable it per-payload without interfering with the package's existing interface, perhaps a new function `WriteUnescaped` could be added, which called another new function `WriteCodeUnescaped`?

**Comment by aeneasr** [comment]
Alternatively we could add the possibility to add options, so like:

```
h.d.Writer(..., herodot.UnescapedJSON())
```

which would not break existing interface and make it easy to specify this. What do you think?

**Comment by fenech** [comment]
Do you mean `Write`, or am I misunderstanding?

**Comment by aeneasr** [comment]
Sorry, yes!

**Comment by oliverpool** [comment]
I guess that this issue can be closed, since #1212 has been merged.
